### PR TITLE
Add a default max_step to processors using the smallest gate time

### LINKED ISSUE #2003: processor.run_state and qc.run returning different states
### Bug Description

Perhaps I am not understanding how a qutip device works.

I have been following the example from https://qutip.org/docs/latest/guide/qip/qip-processor.html#qip-processor as I want to add noise to my quantum circuits.

However, I saw that using ```processor.run_state(init_state)``` returns a different state compared to ```qc.run(init_state)``` for certain circuits.

### Code to Reproduce the Bug

```shell
import numpy as np

from qutip import basis, tensor
from qutip.qip.circuit import QubitCircuit
from qutip.qip.device import LinearSpinChain


num_qubits = 3
init_state = tensor([basis(2,1) for x in range(num_qubits)])
qc = QubitCircuit(num_qubits)
qc.add_gate("CNOT", targets=1, controls=[0])

# Uses a LinearSpinChain processor to obtain the final state
processor = LinearSpinChain(num_qubits)
processor.load_circuit(qc)
processor_state = processor.run_state(init_state).states[-1].tidyup(1.0e-6)

# Obtains state by running the circuit directly
circuit_state = qc.run(init_state)

print(processor_state)
print()
print(circuit_state)
print()

# Calculates the fidelity between both states to show that states are not just differing by a phase
fidelity = np.abs((circuit_state.dag() *  processor_state)[0])

print("Fidelity:", np.abs(fidelity[0]))
```

### Code Output

```shell
Quantum object: dims = [[2, 2, 2], [1, 1, 1]], shape = (8, 1), type = ket
Qobj data =
[[0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [1.]]

Quantum object: dims = [[2, 2, 2], [1, 1, 1]], shape = (8, 1), type = ket
Qobj data =
[[0.]
 [0.]
 [0.]
 [0.]
 [0.]
 [1.]
 [0.]
 [0.]]

Fidelity: [0.]
```

### Expected Behaviour

By just adding an RY gate to the circuit, both methods of obtaining the state now returns the same state with unit fidelity. 

```
import numpy as np

from qutip import basis, tensor
from qutip.qip.circuit import QubitCircuit
from qutip.qip.device import LinearSpinChain


num_qubits = 3
init_state = tensor([basis(2,1) for x in range(num_qubits)])
qc = QubitCircuit(num_qubits)
qc.add_gate("RX", targets=1,arg_value = np.pi/2)
qc.add_gate("CNOT", targets=1, controls=[0])

# Uses a LinearSpinChain processor to obtain the final state
processor = LinearSpinChain(num_qubits)
processor.load_circuit(qc)
processor_state = processor.run_state(init_state).states[-1].tidyup(1.0e-6)

# Obtains state by running the circuit directly
circuit_state = qc.run(init_state)

print(processor_state)
print()
print(circuit_state)
print()

# Calculates the fidelity between both states to show that states are not just differing by a phase
fidelity = np.abs((circuit_state.dag() *  processor_state)[0])

print("Fidelity:", np.abs(fidelity[0]))

```

### Your Environment

```shell
QuTiP: Quantum Toolbox in Python
================================
Copyright (c) QuTiP team 2011 and later.
Current admin team: Alexander Pitchford, Nathan Shammah, Shahnawaz Ahmed, Neill Lambert, Eric Giguère, Boxi Li, Jake Lishman and Simon Cross.
Board members: Daniel Burgarth, Robert Johansson, Anton F. Kockum, Franco Nori and Will Zeng.
Original developers: R. J. Johansson & P. D. Nation.
Previous lead developers: Chris Granade & A. Grimsmo.
Currently developed through wide collaboration. See https://github.com/qutip for details.

QuTiP Version:      4.7.0
Numpy Version:      1.21.5
Scipy Version:      1.7.3
Cython Version:     0.29.30
Matplotlib Version: 3.5.2
Python Version:     3.7.13
Number of CPUs:     8
BLAS Info:          INTEL MKL
OPENMP Installed:   False
INTEL MKL Ext:      False
Platform Info:      Windows (AMD64)
================================================================================
Please cite QuTiP in your publication.
================================================================================
For your convenience a bibtex reference can be easily generated using `qutip.cite()`
```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by BoxiLi**
**Checklist**
Thank you for contributing to QuTiP! Please make sure you have finished the following tasks before opening the PR.

- [x] Include the changelog in a file named: `doc/changes/<PR number>.<type>` 'type' can be one of the following: feature, bugfix, doc, removal, misc, or deprecation (see [here](http://qutip.org/docs/latest/development/contributing.html#Changelog%20Generation) for more information).

**Description**
Add a default max_step to processors using the smallest gate time.

**Related issues or PRs**
fix #2003

### CODE REVIEW

**Comment by BoxiLi** [comment]
Ok, although adding a default `max_step` using the smallest gate time solves the issue. It also significantly increases the computation time in some cases. In particular, for DispersiveCavityQED, the two qubit gate time is ~1000 while the single qubit gate takes only ~1 time unit. Using the `max_step=1` drastically increases the number of time steps needed.

File: qutip/qip/device/processor.py [hunk 1]
```diff
@@ -738,6 +739,11 @@ def run_state(self, init_state=None, analytical=False, states=None,
         else:
             kwargs["c_ops"] = sys_c_ops
 
+        # set the max step size as 1/2 of the smallest gate time.
+        options = kwargs.get("options", Options())
+        if options.max_step == 0.:
+            options.max_step = np.min(np.diff(self.get_full_tlist())) / 22
```
Comment (hodgestar) [inline qutip/qip/device/processor.py:745]: Why 22? :)

**Review by hodgestar — CHANGES_REQUESTED** [review]
The code looks great, I would just like a comment explaining the significance of the number 22. :)

File: qutip/qip/device/processor.py (hunk 1 above)
Comment (BoxiLi) [reply qutip/qip/device/processor.py:745]: Ah, that was a mistake, it should just be `/2`. But still, this does not fully solve the problem above.

**Comment by hodgestar** [comment]
@Ericgig I am wondering whether in v5.1 we should add a way for users to specify the timescale of a QobjEvo. At the moment if one knows the timescale when calling the solver one can just set max step, but in code like Boxi's where one is handling QobjEvos or pulses that were created somewhere else, it can be a lot of extra work to determine what the max_step should be, and it would be good if the QobjEvo already had that information.

The downside is that one might have to specify the time step when creating the QobjEvo, and that might be too annoying or too easy to forget to do.

**Review by hodgestar — APPROVED** [review]
Ah! 2 makes a lot more sense. :)

**Pull request merged**